src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,200 @@
+    pseudo["header"]["has_gipaw"]            && push!(unsupported, "gipaw data")
+    pseudo["header"]["pseudo_type"] == "1/r" && push!(unsupported, "Coulomb")
+    length(unsupported) > 0 && error("Pseudopotential contains the following unsupported" *
+                                     " features/quantities: $unsupported")
+
+    Zion        = Int(pseudo["header"]["z_valence"])     # Pseudo-ion charge
+    rgrid       = pseudo["radial_grid"]                  # Radial grid
+    drgrid      = pseudo["radial_grid_derivative"]       # Integration coeff.s
+    lmax        = pseudo["header"]["l_max"]              # Maximum angular momentum channel
+    vloc        = pseudo["local_potential"] ./ 2         # Local potential (Ry -> Ha)
+    vloc_interp = linear_interpolation((rgrid, ), vloc)  # Interp.or for the local potential
+
+    # Kleinman-Bylander β projectors projs[l+1][i][ir]
+    # NB: UPFs store rβ, not just β
+    projs = map(0:lmax) do l
+        betas_l = filter(beta -> beta["angular_momentum"] == l, pseudo["beta_projectors"])
+        map(beta -> beta["radial_function"] ./ 2, betas_l)  # Ry -> Ha
+    end
+    projs_interp = map(projs) do projs_l
+        map(projs_l) do proj
+            ir_cut = lastindex(proj)
+            linear_interpolation((rgrid[1:ir_cut], ), proj)
+        end
+    end
+
+    # β-projector coupling coefficients h[l+1][i,j] (also called e_{KB}, D_{ij})
+    h = Matrix[]
+    n = 1
+    for l = 0:lmax
+        nproj_l = length(projs[l+1])
+        Dij_l = pseudo["D_ion"][n:n+nproj_l-1, n:n+nproj_l-1] .* 2  # 1/Ry -> 1/Ha
+        push!(h, Dij_l)
+        n += nproj_l
+    end
+
+    # Pseudo-atomic wavefunctions
+    # Currently not used; can be used for initializing the starting wavefunction and as
+    # projectors for projected densities of states, projected wavefunctions, and DFT+U(+V).
+    pswfcs = map(0:lmax - 1) do l
+        pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
+            pswfc["angular_momentum"] == l
+        end
+        map(pswfc -> pswfc["radial_function"], pswfcs_l)
+    end
+    pswfc_occs = map(0:lmax - 1) do l
+        pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
+            pswfc["angular_momentum"] == l
+        end
+        map(pswfc -> pswfc["occupation"], pswfcs_l)
+    end
+
+    # Currently not used; can be used for initializing the starting guess density.
+    rhoatom = pseudo["total_charge_density"]
+
+    # Useful precomputed quantities (q-independent parts of Fourier transform integrands)
+    rvlocpzdr = (rgrid .* vloc .+ Zion) .* drgrid
+    r2projsdr = map(projs) do projs_l
+        map(projs_l) do proj
+            ir_cut = lastindex(proj)
+            rgrid[1:ir_cut] .* proj .* drgrid[1:ir_cut]
+        end
+    end
+
+    PspUpf(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h, pswfcs,
+           pswfc_occs, rhoatom, rvlocpzdr, r2projsdr; identifier)
+end
+
+charge_ionic(psp::PspUpf) = psp.Zion
```

Comment:
move to charge_ionic(psp::NormConservingPsp) = psp.Zion, and add Zion to the list of required fields

Reply:
I disagree. Having required fields is not Julian and should be avoided. Such one-liners don't hurt and give more flexibility.

Reply:
> I disagree. Having required fields is not Julian and should be avoided

Julia just doesn't have a good solution to this kind of abstract interfaces. Having lots of accessor functions is definitely not "julian" either. Personally I'm reluctant to add things to the namespace, but that's not the biggest problem: the biggest problem is that now you need two different names for the same thing, which is confusing. There's absolutely nothing wrong with required fields, just as there's nothing wrong with required methods (at least until julia gets around to implementing proper interfaces, at which point we can revisit)

Reply:
> you need two different names for the same thing

Who says you need two different names? `f(x) = x.f` works fine. 

> There's absolutely nothing wrong with required fields

Well that depends. The issue is you are not just imposing to have the field, but also imposing something on the structure of the field, which can imply built-in assumptions about the data layout. We had a long discussion about this when we were designing AtomsBase and the conclusion was only an interface defined upon functions gives you flexibility in the data layout. We can argue on a case-by-case basis if you want as I agree sometimes requiring a field does not harm, but a function-based interface is the more flexible and general one for sure.